linkerd/tls/src/imp/openssl.rs

```diff
@@ -0,0 +1,295 @@
+use crate::{ClientId, HasNegotiatedProtocol, NegotiatedProtocolRef};
+use linkerd_identity::{ClientConfig, Name, ServerConfig};
+use linkerd_io::{self as io, AsyncRead, AsyncWrite, ErrorKind, PeerAddr, ReadBuf, Result};
+use std::net::SocketAddr;
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use std::str::FromStr;
+use tracing::debug;
+
+#[cfg(feature = "boring-tls")]
+use {
+    boring::{
+        fips, ssl,
+        ssl::{SslAcceptor, SslConnector, SslConnectorBuilder, SslMethod, SslVerifyMode},
+        version,
+        x509::store::X509StoreBuilder,
+    },
+    tokio_boring::SslStream,
+};
+#[cfg(not(feature = "boring-tls"))]
+use {
+    openssl::{
+        fips, ssl,
+        ssl::{Ssl, SslAcceptor, SslConnector, SslConnectorBuilder, SslMethod, SslVerifyMode},
+        version,
+        x509::store::X509StoreBuilder,
+    },
+    tokio_openssl::SslStream,
+};
+
+#[derive(Clone)]
+pub struct TlsConnector(ssl::SslConnector);
+
+impl TlsConnector {
+    #[cfg(not(feature = "boring-tls"))]
+    pub async fn connect<IO>(&self, domain: Name, stream: IO) -> Result<client::TlsStream<IO>>
+    where
+        IO: AsyncRead + AsyncWrite + Unpin,
+    {
+        let ssl = self
+            .0
+            .configure()
+            .unwrap()
+            .into_ssl(domain.as_ref())
+            .unwrap();
```

Comment:
in what cases will these `unwrap`s fail? if they should never fail in normal operation, these should probably be `expect`s with messages explaining why; otherwise, we should return an error in this case.